src/serialize.h

```diff
@@ -371,38 +372,73 @@ template<typename Stream, VarIntMode Mode, typename I>
 void WriteVarInt(Stream& os, I n)
 {
     CheckVarIntMode<Mode, I>();
-    unsigned char tmp[(sizeof(n)*8+6)/7];
-    int len=0;
-    while(true) {
-        tmp[len] = (n & 0x7F) | (len ? 0x80 : 0x00);
-        if (n <= 0x7F)
+    switch(Mode) {
+        case VarIntMode::BITSTREAM: {
+            unsigned char size = (sizeof(n)*8)-std::__countl_zero(n);
+            if (!size) { ser_writedata8(os, 0); break; }
+            unsigned char chunks = ceil(size/7.0);
+            while (chunks > 0) {
+                unsigned char offset = (ceil(size/7.0)-chunks)*7;
+                unsigned char selection = chunks == 1 ? pow(2, (size%7 ? size%7 : 7))-1 : 0x7F;
+                unsigned char data = (n & (I)((I)selection << (I)offset)) >> (I)offset;
```

Comment:
In ec2b2ea56c06bf72e53b0ecfb4065ccf2bd09cb0:

This code looks extremely sketchy. What types do you expect `I` to be? Why are you casting the offsets?

Reply:
Reverted Changes, Thanks!